drivers: udc_dwc2: Optimize Completer/DMA mode check

tmon-nordic opened the pull request:
Allow compiler to optimize away Completer and/or Buffer DMA mode check when all instances are known to use one of the modes. The optimization depends on devicetree ghwcfg2 value being valid, so add assert to verify ghwcfg2 devicetree value against hardware.

tmon-nordic committed:
drivers: udc_dwc2: Optimize Completer/DMA mode check

Allow compiler to optimize away Completer and/or Buffer DMA mode check
when all instances are known to use one of the modes. Optimize away
RxFLvl handler if Completer mode is not needed.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -225,18 +225,62 @@ static void dwc2_wait_for_bit(const struct device *dev,
 	struct udc_dwc2_data *const priv = udc_get_private(dev);
 
-	return IS_ENABLED(CONFIG_UDC_DWC2_DMA) && priv->bufferdma;
+	if (!IS_ENABLED(CONFIG_UDC_DWC2_DMA)) {
+		/* Buffer DMA support not compiled */
+		return false;
+	}
+
+	if (ALL_INSTANCES_SUPPORT_BUFFER_DMA) {
+		/* Buffer DMA is used by all instances */
+		return true;
+	}
+
+	if (!ANY_INSTANCE_SUPPORTS_BUFFER_DMA) {
+		/* Buffer DMA is not supported by any instance */
+		return false;
+	}
+
+	return priv->bufferdma;
```
Not sure what I missed, for what configuraton is `dwc2_in_buffer_dma_mode()` not `!dwc2_in_completer_mode()` ?

tmon-nordic replied:
In order to have to resort to run-time detection `CONFIG_UDC_DWC2_DMA` must be enabled and there have to be  at least two `snps,dwc2` instances: one which supports Internal DMA and one that doesn't.

jfischer-no replied:
Okay, but for the time being, can you change it to return `!dwc2_in_completer_mode()`, or at least add a comment, because how it is now, it is a bit confusing for the code reviewer.

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -2015,6 +2059,10 @@ static int udc_dwc2_init_controller(const struct device *dev)
 	ghwcfg3 = sys_read32((mem_addr_t)&base->ghwcfg3);
 	ghwcfg4 = sys_read32((mem_addr_t)&base->ghwcfg4);
 
+	__ASSERT(ghwcfg2 == config->ghwcfg2,
```
Can you please add checks for 
```c
ghwcfg1 == config->ghwcfg1
ghwcfg4 == config->ghwcfg4
```
as well?

tmon-nordic replied:
Added, and changed it to not actually read the registers at runtime if asserts are not enabled.

tmon-nordic committed:
drivers: udc_dwc2: Use devicetree ghwcfg values

Use devicetree provided ghwcfg values in all contexts, not only in early
initialization. Add asserts to verify devicetree ghwcfg values against
hardware.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -2015,12 +2051,22 @@ static int udc_dwc2_init_controller(const struct device *dev)
 	gsnpsid = sys_read32((mem_addr_t)&base->gsnpsid);
 	priv->wa_essregrestored = gsnpsid < USB_DWC2_GSNPSID_REV_5_00A;
```
> > ```
> > * The values must already be present in Devicetree in order for the driver to work at all, because they are used during pre-init
> > ```
> 
> No they must not, it is a workaround for broken hardware design. The workaround could be redesigned and ghwcfg properties removed.

Back up your claims with working PR.

> > ```
> > * There are assert statements that ensure in debug builds that the DT values match the runtime ones in the IP
> > ```
> 
> ASSERTS does not ensure anything, they are just rarely used helpers during development.

Asserts do ensure that the obvious things are proper. The point of asserts is to be able to take advantage of very basic assumptions that have always to be fulfilled. Failing assert means that the code is wrong. In my opinion these asserts are perfectly placed and doing exactly what should be expected from an assert.

> > Given the above, what exactly is the downside of using values from Devicetree? because the upside is clear: saving both code space and execution time
> 
> What execution time? It only executes once during initialization. There is no reason to have properties in the DT that can be obtained with a single line of code at runtime.

Taking advantage of `ghwcfg` register knowledge at compile time allows compiler to optimize away whole Completer handling when building with DMA enabled. Without `ghwcfg` knowledge this optimization is not possible. This not only cuts off code, but also removes conditional branch at every single USB transfer.

You are directly fighting against one of key Zephyr Distinguishing Features
https://github.com/zephyrproject-rtos/zephyr/blob/b677e828d0750a25a6923a0ec594a9fb06585101/doc/introduction/index.rst?plain=1#L114-L116
